kong/templates/nginx_kong.lua

```diff
@@ -59,16 +59,14 @@ init_worker_by_lua_block {
 
 > if (role == "traditional" or role == "data_plane") and #proxy_listeners > 0 then
 upstream kong_upstream {
+    balancer_by_lua_block {
+        Kong.balancer()
+    }
     server 0.0.0.1;
-
     # injected nginx_upstream_* directives
 > for _, el in ipairs(nginx_upstream_directives) do
     $(el.name) $(el.value);
 > end
-
-    balancer_by_lua_block {
-        Kong.balancer()
-    }
```

Comment:
This change was voluntary in 5c1a7bb96e6aee1579552d1dff06778359e195b6 as part of the work for dynamic upstream keepalive pools. In the currently most probable scenario (in which a user has both a leftover `nginx_upstream_keepalive` property _and_ an `upstream_keepalive_pool_size` one, it is important that the balancer block be defined _after_ the injected `keepalive` directive, in order for it to take precedence over Nginx's stock keepalive module. The warning serves its purpose and indicated to users that they should remove their old, deprecated `nginx_upstream_keepalive` property, and it will go away.

Reply:
@thibaultcha, got it. I force pushed a better fix.

Reply:
@thibaultcha Is there a difference in user-visible behavior with the change of defaults produced in the latest version of this PR (de6698c)? (I mean, other than fixing broken cases that dynamic upstream keepalive pools fix, of course)

Being a change of defaults, without a deeper understanding of what's going on here, I would assume so, but I'm unclear if it's the case. If that's the case, would this this be a breaking change (people upgrading without touching their configs would end up with different behavior), and thus to be done only in 3.0, right?

Reply:
Whichever of `balancer_by_lua` or Nginx's stock `keepalive` directive is defined _last_ will take effect. If both `balancer_by_lua` (with the keepalive patch) _and_ the `keepalive` directive are defined (in whichever order), then the warning will be shown. Once again, the goal was that in this case, the dynamic pool keepalive would take precedence.

By keeping these defaults, I aimed at making it easier for users to revert to the stock upstream keepalive in case the new one had any issue we weren't aware of at time of the release. It's a slower roll-out of the feature with a warning for users to fix their configuration (avoid having both properties set will log both a configuration warning from our conf loader _and_ this balancer warning from Nginx). It's also worth noting that many production deployments probably have an overriden value for this setting, other than the default, and so the warnings probably show up regardless of any work we do to optimize default configuration setups.

By removing the defaults like in the above commit, this means that the warning goes away but that reverting to the stock keepalive module requires `upstream_keepalive_pool_size = 0` _and_ `nginx_upstream_keepalive > 0`. I think that's fine too.